crypto/self_test_core.c

```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/self_test.h>
+
+static OSSL_CALLBACK *ossl_self_test_cb = NULL;
```

Comment:
OPENSSL_CTX? Will we want to be able to set a callback in one OPENSSL_CTX but not another? This sets one callback globally. We've been trying to move away from global data.

Reply:
That stuff hurts my brain..
You mean use OPENSSL_CTX_METHOD and openssl_ctx_get_data()

Reply:
So, err, that takes us back to callback dispatching on a per OPENSSL_CTX basis, doesn't it?  I.e. more or less my original OSSL_CALLBACK proposal (#10390) :wink:

Reply:
Not sure #10390 is really relevant here.
@mattcaswell How do I move this forward? This PR has been kicking around for a few months now :)

Reply:
> You mean use OPENSSL_CTX_METHOD and openssl_ctx_get_data()?

Yes, exactly that.

Reply:
Yeah I tried doing that, but this just is not going to work per libctx for self tests..

The fipinstall application does not know what its libctx is as it loads via configuration - and it needs to set the callback before doing that. i.e it just calls CONF_modules_load()

Reply:
So in that case it just passes NULL through for the libctx in order to get the default. The function that implements setting the callback will then resolve that to the "real" libctx via `openssl_ctx_get_concrete`

Reply:
There is no default libctx associated with the fips provider...

Reply:
The FIPS install applications does not run inside the FIPS provider. It links against libcrypto which does have a default context.

Reply:
Fair enough.. But the callback gets called from the fips provider for the install.
How would it know which libctx to use in order to find the callback in the exdata (because it wont be the fips libctx). Yeah we could use NULL here too - but is that what we want?
A single callback associated with the default libctx? Storing the libctx in the ctx_method struct isnt going to help for this case..

Reply:
I think a single callback that gets passed the libctx is probably a better way to handle this (which it already does). This callback is really meant for diagnostic testing.

Reply:
> How would it know which libctx to use in order to find the callback in the exdata

The set and get callback functions are modified to take a libctx as a parameter:

```` C
void OSSL_SELF_TEST_set_callback(OPENSSL_CTX *libctx, OSSL_CALLBACK *cb);
OSSL_CALLBACK *OSSL_SELF_TEST_get_callback(OPENSSL_CTX *libctx);
````

The application calls OSSL_SELF_TEST_set_callback() and passes the libctx that should be used (or NULL for the default libctx).

The implementation of OSSL_SELF_TEST_set_callback() stores the provided callback function pointer in the libctx that it has been passed.

The implementation of OSSL_SELF_TEST_get_callback() looks up the callback to be used in the libctx and returns it.

fipsprov.c calls OSSL_get_self_test_cb() which just calls OSSL_SELF_TEST_get_callback() under the covers. It passes as an argument the result of calling c_get_libctx, i.e. something like this:

    OSSL_self_test_cb_fn *fn = OSSL_get_self_test_cb(in);

    selftest_params.event_cb = (fn != NULL) ? fn(c_get_libctx(provider)) : NULL;

Reply:
I was doing this locally except for the fn(c_get_libctx(provider))..
So the 'provider' pointer in this case is the fips provider (which just happens to have a NULL libctx during the self test - which sounds like it works by accident).

Reply:
I have updated the PR to include the OPENSSL_CTX